### DISCUSSION TITLE
A certain combination of operations causes 2-3 minute exports

### DISCUSSION THREAD

**Discussion by jawngee** [libvips/libvips#3982] [General]
**Describe the bug**

We are redacting faces from images by obscuring them with blurring and/or pixelation.

If we do just blur or we do just pixelation, the whole export finishes within a second.

If we do both, the whole export takes several minutes.

I don't know the internals of libvips, but I'm assuming it doesn't actually perform all of the requested operations until export, so I don't believe it's the export itself that is slow.

This issue only seems to appear in a Docker image.

**To Reproduce**

1. Blur the source image
2. Pixelate source image by resizing it to 1.0/n where n is the desired pixelation and then resize it back using nearest neighbor
3. Export to any file type

This is what the go code looks like:

```go
_ = redactedParts.GaussianBlur(*redact.Blur)
_ = redactedParts.Resize(1.0 / *redact.Pixelate, vips.KernelLanczos3)
_ = redactedParts.Resize(*redact.Pixelate, vips.KernelNearest)
```
The `GaussianBlur()` is calling `sharpen_image()` under the covers and the `Resize()` is calling `resize_image()`

The source images are 3840x2560 but that doesn't seem to have any bearing as the same delay happens with much smaller images.

**Expected behavior**
For it not to take 3 minutes.

**Environment**
 - OS: Debian Bookworm Slim
 - Vips: 8.14 (whatever the latest debian bookworm package version is)
 - Govips latest

**Additional context**
Of course this issue doesn't exist when running locally on macOS during development.  It's only when running in a docker image that the issue pops up because that's the way the universe seemingly operates.  

The base image for the Dockerfile looks like:

```Dockerfile
FROM debian:bookworm-slim as foxyvips
RUN apt update \
    && apt install -y  \
    fontconfig \
    libde265-0 \
    libheif1 \
    libvips
```
Maybe I'm missing a package?

**Comment by jawngee** [comment]
I've created a small reproduction:

https://github.com/jawngee/slow-vips-ops-repro/tree/main

You can run it using `docker compose up`

I've tested it on a couple of different machines to the same effect.

**Comment by lovell** [comment]
This might be a question for https://github.com/davidbyttow/govips e.g. https://github.com/davidbyttow/govips/issues/289 will be preventing certain optimisations/caching from occurring and https://github.com/davidbyttow/govips/issues/197 will be required to switch from sequential to random access - you may find the latter more appropriate given the large Gaussian blur operation.

**Comment by jcupitt** [comment]
As well as Lovell's excellent advice, you can speed your code up a bit with some small refactoring.

I made a tiny test framework in python:

```python
#!/usr/bin/env python3

import sys
import pyvips
import timeit

def blur(filename):
    image = pyvips.Image.new_from_file(filename, access="sequential")
    image = image.gaussblur(18)
    data = image.write_to_buffer(".jpg", Q=100)

def pix(filename):
    image = pyvips.Image.new_from_file(filename, access="sequential")
    image = image.resize(1.0/16).resize(16)
    data = image.write_to_buffer(".jpg", Q=100)
    
def pix_and_blur(filename):
    image = pyvips.Image.new_from_file(filename, access="sequential")
    image = image.gaussblur(18)
    image = image.resize(1.0/16).resize(16)
    data = image.write_to_buffer(".jpg", Q=100)

def time(statement):
    t = timeit.timeit(statement, number=5, globals=globals())
    print(f"{statement} took {t:.1f}s")

time("blur(sys.argv[1])")
time("pix(sys.argv[1])") 
time("pix_and_blur(sys.argv[1])") 
```

Which I think matches your dockerfile. With a debug build of libvips I see:

```
$ ./pixelate.py data/test.jpg 
blur(sys.argv[1]) took 0.76s
vips_image_dispose: temp-16 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-49 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-82 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-115 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-148 vips_convi_uchar_vector_gen computed 146%
pix(sys.argv[1]) took 0.83s
vips_image_dispose: temp-181 vips_reduceh_uchar_vector_gen computed 247%
vips_image_dispose: temp-182 vips_shrinkh_gen computed 359%
vips_image_dispose: temp-221 vips_reduceh_uchar_vector_gen computed 247%
vips_image_dispose: temp-222 vips_shrinkh_gen computed 359%
vips_image_dispose: temp-260 vips_reduceh_uchar_vector_gen computed 247%
vips_image_dispose: temp-261 vips_shrinkh_gen computed 359%
vips_image_dispose: temp-299 vips_reduceh_uchar_vector_gen computed 247%
vips_image_dispose: temp-300 vips_shrinkh_gen computed 359%
vips_image_dispose: temp-338 vips_reduceh_uchar_vector_gen computed 247%
vips_image_dispose: temp-339 vips_shrinkh_gen computed 359%
vips_image_dispose: temp-404 vips_reduceh_uchar_vector_gen computed 247%
vips_image_dispose: temp-405 vips_shrinkh_gen computed 359%
vips_image_dispose: temp-379 vips_convi_uchar_vector_gen computed 880%
vips_image_dispose: temp-468 vips_reduceh_uchar_vector_gen computed 247%
vips_image_dispose: temp-469 vips_shrinkh_gen computed 359%
vips_image_dispose: temp-443 vips_convi_uchar_vector_gen computed 880%
pix_and_blur(sys.argv[1]) took 51.16s
```

The "computed 880%" etc. is a debug feature of libvips which counts pixels that pass through operations compared to the expected number of pixels in an image -- you can see that running a large radius gaussblur before a resize causes almost 9x overcomputation of earlier stages.

I'd first swap the two resizes for shrink and zoom:

```
image = image.shrink(16, 16).zoom(16, 16)
```

These are low level operations that do block averaging and block zooming.

https://www.libvips.org/API/current/libvips-resample.html#vips-shrink

https://www.libvips.org/API/current/libvips-conversion.html#vips-zoom

I see:

```
$ ./pixelate.py data/test.jpg 
blur(sys.argv[1]) took 0.8s
vips_image_dispose: temp-16 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-49 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-82 vips_convi_uchar_vector_gen computed 146%
pix(sys.argv[1]) took 0.3s
vips_image_dispose: temp-115 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-148 vips_convi_uchar_vector_gen computed 146%
vips_image_dispose: temp-266 vips_convi_uchar_vector_gen computed 489%
pix_and_blur(sys.argv[1]) took 30.7s
```

For the pathological case (blur then shrink then zoom) you have a few choices. The simplest is to cache after the gaussblur, eg.:

```
def pix_and_blur(filename):
    image = pyvips.Image.new_from_file(filename, access="sequential")
    image = image.gaussblur(18)
    image = image.copy_memory().shrink(16, 16).zoom(16, 16)
    data = image.write_to_buffer(".jpg", Q=100)
```

Will render the blur to memory, then do the shrink and zoom. This gets the time down to 0.9s, but it will use quite a bit of memory for the cache.

You can use random access mode (as Lovell said):

```python
def pix_and_blur(filename):
    image = pyvips.Image.new_from_file(filename)
    image = image.gaussblur(18)
    image = image.shrink(16, 16).zoom(16, 16)
    data = image.write_to_buffer(".jpg", Q=100)
```

Gives 1.7s, but again will raise memory or maybe (for very large images) disc use.

**Reply by jcupitt** [reply]
And of course another option is to not do both blur and pixelate. I'd think just one or the other would be fine.

**Comment by jcupitt** [comment]
Just for fun I made a tiny thing to blur a circular area of an image:

```python
#!/usr/bin/env python3

import sys
import pyvips

def blur(filename, x, y, r):
    """blur a circular area at (x, y) radius r.
    """
    image = pyvips.Image.new_from_file(filename)

    # we blur the whole image, make a mask for the circle, then use ifthenelse
    # to pick pixels from either the blurred version or the original

    blur = image.gaussblur(18)

    # bounding box of the circle
    circle_image = image.crop(x - r, y - r, r * 2, r * 2)
    circle_blur = blur.crop(x - r, y - r, r * 2, r * 2)

    # make an ideal circular mask
    mask = pyvips.Image.mask_ideal(r * 2, r * 2, 1,
                                   uchar=True, reject=True, optical=True)
    
    # and mask pixels ... we could soften the edge, but maybe a hard edge
    # adtuically looks nicer
    masked = mask.ifthenelse(circle_blur, circle_image)

    # and insert the masked image back into the original
    image = image.insert(masked, x - r, y - r)
    
    return image

image = blur(sys.argv[1], int(sys.argv[3]), int(sys.argv[4]),int(sys.argv[5]))
image.write_to_file(sys.argv[2])
```

With this test image:

![136944814-5635fb35-2e9d-454a-99be-0bc3acf446d7](https://github.com/libvips/libvips/assets/580843/5c86fe50-e6f1-4e8f-9de0-6f9e9905e69e)

I see:

```
$ time ./pixelate.py ~/pics/136944814-5635fb35-2e9d-454a-99be-0bc3acf446d7.jpg x.jpg 700 600 200

real	0m0.173s
user	0m0.372s
sys	0m0.070s
```

To make:

![x](https://github.com/libvips/libvips/assets/580843/4e335255-b173-495a-b360-b0baed71e5bd)

**Reply by jcupitt** [reply]
I meant to say, blurring the whole image and then cropping a box from it sounds very slow, but because libvips is demand-driven it'll only compute the bit you actually use in the output. It's a simple way to avoid having to worry about abyss handling.

You could swap the gaussblur for shrink/zoom if you want to pixelate instead. 

This won't handle circles which go outside the image area -- if that's important, I'd use `gravity` to grow the original image out by `r` in all directions, then do the thing above, then crop again at the end.
